src/ImageSharp/Formats/Tiff/TiffDecoderCore.cs

```diff
@@ -509,6 +473,299 @@ private void DecodeStripsChunky<TPixel>(ImageFrame<TPixel> frame, int rowsPerStr
+            this.YcbcrSubSampling,
+            this.byteOrder);
+
+    private TiffBaseDecompressor CreateDecompressor<TPixel>(int frameWidth, int bitsPerPixel)
+        where TPixel : unmanaged, IPixel<TPixel> =>
+        TiffDecompressorsFactory.Create(
+            this.Options,
+            this.CompressionType,
+            this.memoryAllocator,
+            this.PhotometricInterpretation,
+            frameWidth,
+            bitsPerPixel,
+            this.ColorType,
+            this.Predictor,
+            this.FaxCompressionOptions,
+            this.JpegTables,
+            this.OldJpegCompressionStartOfImageMarker.GetValueOrDefault(),
+            this.FillOrder,
+            this.byteOrder);
+
+    private IMemoryOwner<ulong> ConvertNumbers(Array array, out Span<ulong> span)
+    {
+        if (array is Number[] numbers)
+        {
+            IMemoryOwner<ulong> memory = this.memoryAllocator.Allocate<ulong>(numbers.Length);
+            span = memory.GetSpan();
+            for (int i = 0; i < numbers.Length; i++)
+            {
+                span[i] = (uint)numbers[i];
+            }
+
+            return memory;
+        }
+
+        DebugGuard.IsTrue(array is ulong[], $"Expected {nameof(UInt64)} array.");
+        span = (ulong[])array;
+        return null;
+    }
+
+    /// <summary>
+    /// Calculates the size (in bytes) for a pixel buffer using the determined color format.
+    /// </summary>
+    /// <param name="width">The width for the desired pixel buffer.</param>
+    /// <param name="height">The height for the desired pixel buffer.</param>
+    /// <param name="plane">The index of the plane for planar image configuration (or zero for chunky).</param>
+    /// <returns>The size (in bytes) of the required pixel buffer.</returns>
+    private int CalculateStripBufferSize(int width, int height, int plane = -1)
+    {
+        DebugGuard.MustBeLessThanOrEqualTo(plane, 3, nameof(plane));
+
+        int bitsPerPixel = 0;
+
+        if (this.PlanarConfiguration == TiffPlanarConfiguration.Chunky)
+        {
+            DebugGuard.IsTrue(plane == -1, "Expected Chunky planar.");
+            bitsPerPixel = this.BitsPerPixel;
+        }
+        else
+        {
+            switch (plane)
+            {
+                case 0:
+                    bitsPerPixel = this.BitsPerSample.Channel0;
+                    break;
+                case 1:
+                    bitsPerPixel = this.BitsPerSample.Channel1;
+                    break;
+                case 2:
+                    bitsPerPixel = this.BitsPerSample.Channel2;
+                    break;
+                case 3:
+                    bitsPerPixel = this.BitsPerSample.Channel2;
+                    break;
+                default:
+                    TiffThrowHelper.ThrowNotSupported("More then 4 color channels are not supported");
+                    break;
+            }
+        }
+
+        int bytesPerRow = ((width * bitsPerPixel) + 7) / 8;
```

Comment:
Could also use a helper-method for this, which has all the ugly casts in it, so the call-sites remain clean and descriptive.